storm-core/src/jvm/org/apache/storm/assignments/AssignmentDistributionService.java

```diff
@@ -0,0 +1,232 @@
+            this.service.submit(new DistributeTask(this, i));
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        this.active = false;
+        this.service.shutdownNow();
+        try {
+            this.service.awaitTermination(10l, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            LOG.error("Failed to close assignments distribute service");
+        }
+        this.assignmentsQueue = null;
+    }
+
+    public void addAssignmentsForNode(String node, SupervisorAssignments assignments) {
+        try {
+            boolean success = nextQueue().offer(NodeAssignments.getInstance(node, assignments), 5l, TimeUnit.SECONDS);
+            if (!success) {
+                LOG.warn("Discard an assignment distribution for node {} because the target sub queue is full", node);
+            }
+
+        } catch (InterruptedException e) {
+            LOG.error("Add node assignments interrupted: {}", e.getMessage());
+            throw new RuntimeException(e);
+        }
+    }
+
+    static class NodeAssignments {
+        private String node;
+        private SupervisorAssignments assignments;
+
+        private NodeAssignments(String node, SupervisorAssignments assignments) {
+            this.node = node;
+            this.assignments = assignments;
+        }
+
+        public static NodeAssignments getInstance(String node, SupervisorAssignments assignments) {
+            return new NodeAssignments(node, assignments);
+        }
+
+        public String getNode() {
+            return this.node;
+        }
+
+        public SupervisorAssignments getAssignments() {
+            return this.assignments;
+        }
+
+    }
+
+    /**
+     * Task to distribute assignments.
+     */
+    static class DistributeTask implements Runnable {
+        private AssignmentDistributionService service;
+        private Integer queueIndex;
+
+        DistributeTask(AssignmentDistributionService service, Integer index) {
+            this.service = service;
+            this.queueIndex = index;
+        }
+
+        @Override
+        public void run() {
+            while (true) {
+                try {
+                    NodeAssignments nodeAssignments = this.service.nextAssignments(queueIndex);
+                    sendAssignmentsToNode(nodeAssignments);
+                } catch (InterruptedException e) {
+                    if (service.isActive()) {
+                        LOG.error("Get an unexpected interrupt when distributing assignments to node, {}", e.getCause());
+                    } else {
+                        // service is off now just interrupt it.
+                        Thread.currentThread().interrupt();
+                    }
+                }
+            }
+        }
+
+        private void sendAssignmentsToNode(NodeAssignments assignments) {
+            SupervisorClient client = SupervisorClient.getConfiguredClient(service.getConf(), assignments.getNode());
```

Comment:
Can we do something where if the send is too old that we don't do it?

Because we have 2 ways to get this information to the client I can see us having race conditions where a supervisor sees State A followed by State B and then back again to State A.  I would prefer to avoid something like that, and timeout the send if it is more than 10 seconds old, because the supervisor should be checking every 10 seconds anyways.